test-runner-frontend/src/Views/Dashboard.js

```diff
@@ -50,290 +47,224 @@ export default function Dashboard() {
+      improved: improvementsAndRegressions
+        ?.filter((improved) => improved === 'improved')
+        .reduce((first, second) => first + 1, counter),
+      regressed: improvementsAndRegressions
+        ?.filter((regressed) => regressed === 'regressed')
+        .reduce((first, second) => first + 1, counter),
+    };
 
-  const [testRunSelection, setTestRunSelection] = useState('All Test Runs');
-  const [machineSelection, setMachineSelection] = useState('All Machines');
+    return improvedVsRegressed;
+  };
+
+  const flattenedSortedTestRuns = (sortedTestRuns) => {
+    return sortedTestRuns.map((testRunsSortedByMachine) => {
+      return testRunsSortedByMachine.map((testRun) => testRun).filter((testRun) => testRun);
+    });
+  };
+
+  const getAllBenchmarks = (benchmarksByMachine) => {
+    const allBenchmarks = benchmarksByMachine.flatMap((benchmarkByMachine) => {
+      return benchmarkByMachine
+        .map((testRun) => {
+          return testRun;
+        })
+        .filter((testRun) => testRun);
+    });
+
+    return allBenchmarks;
+  };
+
+  const getBenchmarksByMachine = (flattenedTestRuns) => {
+    const benchmarksByMachine = flattenedTestRuns.map((testRunsSortedByMachine) => {
+      return testRunsSortedByMachine
+        .map((testRun, index) => {
+          if (testRunsSortedByMachine[index + 1] && testRunsSortedByMachine[index + 1].statistics) {
+            const getImprovedVsRegressedValues = generateComparisonBetween(
+              testRun,
+              testRunsSortedByMachine[index + 1] ? testRunsSortedByMachine[index + 1] : {}
+            );
+            testRun.improvedVsRegressed.improved = getImprovedVsRegressedValues?.improved
+              ? getImprovedVsRegressedValues.improved
+              : 0;
+            testRun.improvedVsRegressed.regressed = getImprovedVsRegressedValues?.regressed
+              ? getImprovedVsRegressedValues.regressed
+              : 0;
+          }
+          return {
+            id: testRun.id,
+            commitHash: testRun.commitHash,
+            branch: testRun.branch,
+            status: testRun.status,
+            benchmarks: testRun.benchmarks,
+            machine: testRun.machine,
+            date: testRun.date,
+            statistics: testRun.statistics,
+            improvedVsRegressed: testRun.improvedVsRegressed,
+          };
+        })
+        .filter((testRun) => testRun);
+    });
+    return benchmarksByMachine;
+  };
+
+  const compareBenchmarks = (testRunResults) => {
+    const machines = getMachineNames(testRunResults);
+    const sortedTestRuns = sortTestRunsByMachine(machines, testRunResults);
+    const flattenedTestRuns = flattenedSortedTestRuns(sortedTestRuns);
+    const benchmarksByMachine = getBenchmarksByMachine(flattenedTestRuns);
+    const allBenchmarks = getAllBenchmarks(benchmarksByMachine);
+    setTestRunResultsCopy(allBenchmarks);
+    setTestRunResults(allBenchmarks);
+    setLoadingState(false);
+  };
```

Comment:
there are a lot of functions here. can we please move these into a utils file and then we can independently test them.